dynawo/sources/Util/DYNIoDicos.cpp

```diff
@@ -38,44 +38,59 @@ using std::vector;
 
 namespace DYN {
 
-static bool readLine(string& line, string& key, string& phrase);
+IoDicos::IoDicos() {
+}
 
-shared_ptr<IoDicos> IoDicos::getInstance() {
-  static shared_ptr<IoDicos> instance;
-  if (instance)
-    return instance;
+IoDicos::~IoDicos() {
+}
 
-  instance.reset(new IoDicos());
+IoDicos& IoDicos::instance() {
+  static IoDicos instance;
   return instance;
 }
 
 void IoDicos::addPath(const string& path) {
+  instance().addPath_(path);
+}
+
+void IoDicos::addPath_(const string& path) {
   paths_.push_back(path);
 }
 
 bool IoDicos::hasIoDico(const string& dicoName) {
-  return ( getInstance()->dicos_.find(dicoName) != getInstance()->dicos_.end());
+  return instance().hasIoDico_(dicoName);
+}
+
+bool IoDicos::hasIoDico_(const string& dicoName) {
+  return (dicos_.find(dicoName) != dicos_.end());
 }
 
-boost::shared_ptr<IoDico> IoDicos::getIoDico(const string& dicoName) {
-  if (hasIoDico(dicoName)) {
-    return getInstance()->dicos_[dicoName];
+const IoDico& IoDicos::getIoDico(const string& dicoName) {
+  return instance().getIoDico_(dicoName);
+}
+
+const IoDico& IoDicos::getIoDico_(const string& dicoName) {
+  if (hasIoDico_(dicoName)) {
+    return *dicos_[dicoName];
   } else {
     throw MessageError("Unknown dictionary '" + dicoName + "'");
   }
 }
 
+IoDico& IoDicos::getNonCstIoDico_(const string& dicoName) {
+  return const_cast<IoDico&>(getIoDico_(dicoName));
```

Comment:
the use of const cast is not recommenced.The getIoDico_ method should have a non const version

Reply:
why not recommended ? I see nothing here https://google.github.io/styleguide/cppguide.html#Use_of_const

Reply:
it is never good to use a const_cast :any good book on C++ will say that using a const_cast is generally the reflect of an issue in design

Reply:
@rosiereflo If you read Effective C++ by Scott Meyers, you will see that there are some use cases where it's totally valid to use `const_cast`. For instance, to avoid duplication in non trivial getters, Scott Meyers explain that it's allowed to use `const_cast` in the non-const getter to call the const version:

1) at the beginning, `this` is not const
2) you cast `this` to `const*` --> OK because you are more restrictive
3) you call the getter to get an attribute, which is const
4) you cast the attribute to a non-const --> OK because at the beginning you weren't const.

Here you have to chose between:
- duplicate a 10 lines method
- allow to use const_cast for this specific case.

Personally, in powsybl-iidm4cpp, I choose to not duplicate because duplicate code is error prone: a developer could change one of the two getters.

http://www.richelbilderbeek.nl/CppAvoidDuplicationInConstAndNonConstMemberFunctions.htm

Reply:
Yes this was the book I was thinking to when writing my message (even if I disagree with 3 and 4). Why would we have a const getter and then a const_cast and not the opposite?

Reply:
btw I agree with the fact that duplicating is not good, but in this case I want to make sure that this cannot be avoided for this class in dynawo